### ISSUE TITLE
PersistentParallelExecutor grain-size threshold not effective for ViT-Base CPU train (was: #313 dispatch-cost dominance)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#319]
## Summary

Issue #313 was closed by [#316](https://github.com/ooples/AiDotNet.Tensors/pull/316) ("grain-size serial-fallback for PersistentParallelExecutor"), but the empirical bottleneck it filed against still reproduces unchanged on AiDotNet.Tensors 0.75.3. Filing as a separate issue per maintainer request rather than reopening since the original was closed against a specific PR that did land.

## What persists post-#316

`PersistentParallelExecutor` thread-coordination overhead still dominates wall-clock on a representative ViT-Base CPU train step.

## Reproduction

Consumer harness `tools/ClipPerfHarness/Program.cs vit` mode in `ooples/AiDotNet` PR #1224. ViT-Base config (Dosovitskiy et al. 2021 paper-default): 224×224×3 input, 16-px patches, hidden=768, 12 layers, 12 heads, mlp=3072, 86M params.

```bash
dotnet run -c Release --project tools/ClipPerfHarness -- vit
```

Hardware: consumer x64 with AVX2 (no GPU). Tensors 0.75.3, .NET 10.0.

## Timing (0.75.3, no profiler)

```
[bench] sub-phase break-down (10 reps each)
  predict x10: 7097 ms total (709.7 ms/iter)
  train   x10: 41832 ms total (4183.2 ms/iter)
  ==> backward+optimizer estimate: 3473 ms/iter
```

## Comparison across versions (same harness, same hardware)

| Tensors version | ViT-Base Train ms/iter | Predict ms/iter |
|---|---:|---:|
| 0.72.0 | 4144 | 670 |
| 0.75.2 + user-side MHA QKV-output FusedLinear fusion | 3864 | 661 |
| **0.75.3 + same fusion (#316 grain-size landed)** | **4183** | **710** |

The 0.75.3 numbers are within run-to-run noise of 0.75.2 — **no observable improvement from #316**.

## Failing consumer test

```
$ dotnet test --filter "VisionTransformerTests.MoreData_ShouldNotDegrade"
[xUnit.net 00:02:10.72]  AiDotNet.Tests.ModelFamilyTests.NeuralNetworks.VisionTransformerTests.MoreData_ShouldNotDegrade [FAIL]
   Error Message: Test execution timed out after 120000 milliseconds
```

`MoreData_ShouldNotDegrade` runs `MoreDataLongIterations = 200` Train iters × 4183 ms/iter ≈ 836 sec, well over the 120 sec xUnit budget. Same as on 0.72.0 / 0.75.2.

## Hypothesis on why #316 doesn't kick in

Two main candidates:

1. **Default grain-size threshold is too high for typical user-level ops.** ViT-Base on a single-batch CPU forward dispatches:
   - QKV projection matmuls: `[1×197, 768] × [768, 768]` = ~150 K elements per matmul
   - Per-head QK^T: `[197, 64] × [64, 197]` = ~12 K per head per layer
   - LayerNorm reductions: 197 × 768 ≈ 150 K elements per layer
   - Attention softmax: ~38 K elements per layer
   - MLP fc1: `[197, 768] × [768, 3072]` = ~600 K per layer
   - Bias adds, broadcasts, residual sums: 1–150 K each

   At fp64 with PyTorch's default `at::internal::GRAIN_SIZE = 32 * 1024` (32 K) elements, *most* of the small ops (LayerNorm reductions, bias adds, attention softmax tiles) sit below the threshold and *should* run serial. If 0.75.3's default is set lower than 32 K, those ops fan out anyway. If it's set higher than 600 K, even the matmuls run serial — which would help peak compute but may also miss real parallelism.

2. **Tape-aware path bypasses the threshold check.** `TrainWithTape` records every op into the gradient tape. The recorded ops may dispatch through a different code path (e.g., a tape-replay executor) that hasn't been updated to consult the new grain-size value. Forward (`Predict`) saw 6.8 % regression noise; train (which heavily exercises the tape) saw 8.3 % regression noise — neither showing the >2× speedup that a working serial-fallback would deliver on this op-size distribution.

## Suggested investigation

- Profile an isolated `Engine.TensorMatMul([1, 768] × [768, 768])` call on 0.75.3 with `dotnet-trace`. If `LowLevelLifoSemaphore.WaitForSignal` is still > 50 % exclusive, the threshold isn't kicking in for this op size.
- Confirm what numerical value the new default grain size takes. If it's a large fraction of typical ViT-Base ops, lower it.
- Audit which kernel paths consult `PersistentParallelExecutor.For(begin, end, body)` vs the new `For(begin, end, grainSize, body)`. Any callers still on the old overload silently get default-parallel behavior.

## What would close this issue

ViT-Base CPU train wall-clock drops to **a few hundred ms / iter** on the same harness — i.e. the actual matmul compute (4.77 % exclusive in the original profile) becomes the dominant cost rather than `LowLevelLifoSemaphore.WaitForSignal` (42.59 % exclusive). PyTorch reference for this class of CPU is 30–80 ms / iter at fp32 — even fp64 should fit in ~150–300 ms / iter once dispatch overhead is amortized.

`VisionTransformerTests.MoreData_ShouldNotDegrade` (and the rest of the consumer-PR B11 timeout cluster — 45 tests) fits in the 120 sec xUnit budget.

This was filed in #313 with [comment 4410979414](https://github.com/ooples/AiDotNet.Tensors/issues/313#issuecomment-4410979414) showing the persistence; raising as a fresh issue for visibility.

**Comment by ooples** [comment]
## Reality check from the maintainer's diagnostic test

The maintainer added [`Issue319TransformerBlockPerfTests`](https://github.com/ooples/AiDotNet.Tensors/blob/fix/319-grain-size-comprehensive-migration/tests/AiDotNet.Tensors.Tests/Engines/Issue319TransformerBlockPerfTests.cs) on PR #321 that runs the actual ViT-Base block hot path (12 layers × matmul + bias + GELU + LayerNorm). Re-ran it on top of my LayerNorm migration:

```
Per-op µs/call at ViT-Base shape [197, 768] (over 360 calls each):
   TensorMatMul       :  50697.2 µs/call (total 18251 ms)   ← 80% of block time
   GELU               :   2405.6 µs/call (total  866 ms)
   TensorBroadcastAdd :    527.8 µs/call (total  190 ms)
   LayerNorm          :    255.6 µs/call (total   92 ms)
 Full block chain     : 630.13 ms/iter (over 30 iters)
```

**The dominant cost is TensorMatMul** at ~50 ms/call. With 360 calls per block × 12 blocks per iter = 4,320 matmul calls → ~219 sec just in matmul, even with zero Parallel.For overhead anywhere else.

That matches the user-reported 4183 ms/iter ÷ 12 layers ≈ 350 ms/layer ≈ matmul-dominated.

**The original profile** (`LowLevelLifoSemaphore.WaitForSignal` 42.59% exclusive) was workers idle-waiting between matmul tiles inside `SimdGemm.DgemmAvx2` — when each tile is small (64×64 doubles), workers wake → do micro-work → block on the next tile boundary → semaphore wait dominates the SAMPLE distribution. But on wall-clock the actual time is in the FMA work, not the waits. The profile's "90% wait" was misleading on this op-class because the sampler counts blocked threads.

## What this means for #319

The Parallel.For comprehensive migration helps secondary ops (LayerNorm, broadcast adds, element-wise) but **does not move the ViT-Base wall-clock budget**. To fit ViT-Base train in xUnit's 120s budget, the lever is `TensorMatMul` kernel throughput:

- BLAS for double — currently `SimdGemm.Dgemm` ships an in-house AVX2 path that maxes ~12 GFLOPS, vs OpenBLAS/MKL ~75 GFLOPS for Dgemm. The CpuEngine.TensorMatMul2D fast path at line 10026 already calls `MatrixMultiplyHelper.TryGemm` but it returns false for double (BlasProvider double-disabled). Wiring up double BLAS would be a 6× kernel speedup and the single biggest #319-class win.
- Larger SimdGemm tile sizes — current 64×64 tiles undersaturate the worker pool's effective parallelism. Larger tiles amortize dispatch.
- Packed-B panel layout for cache efficiency.

The grain-size migration in #316/#321 still has value (LayerNorm goes from 255 µs/call serial-fallback to bare ops, eliminating ~45ms/iter of Parallel.For overhead across the chain). But it's secondary to the matmul kernel.

## What I've contributed

Pushed [LayerNorm migration to PR #321](https://github.com/ooples/AiDotNet.Tensors/pull/321/commits/6e4c8bb6) — both float64 fast path and generic-T fallback go serial inline below 32K elementwise grain. Verified 33/33 LayerNorm tests pass on net10.0 + net471.

The user-facing B11 timeout cluster will need either (a) the Dgemm-via-BLAS work as a follow-up #319-extension, or (b) a smaller-ViT-config workaround in consumer-side test scaffolds. Happy to take on either; the maintainer's diagnostic perf test is the right fixture for measuring impact.

**Comment by ooples** [comment]
## Update: BLAS-default-on commit pushed to PR #321 — measured impact

Pushed [`ddbe6d57`](https://github.com/ooples/AiDotNet.Tensors/commit/ddbe6d57) to PR #321 — flips `BlasProvider.ReadOptIn()` from opt-IN to opt-OUT. Industry-standard contract now matches PyTorch / NumPy / TensorFlow: BLAS is on when `libopenblas` is loadable; opt-out via `AIDOTNET_USE_BLAS=0`. Since `AiDotNet.Native.OpenBLAS` is a transitive NuGet dependency of every consumer install, this engages BLAS for every existing user without any consumer-side action.

### Measured end-to-end on consumer ViT-Base

`tools/ClipPerfHarness -- vit` (full ViT-Base 224×224×3 train step) on the same hardware:

| Tensors version | ViT-Base train ms/iter | Predict ms/iter |
|---|---:|---:|
| 0.75.3 (pre-flip, BLAS off) | 4144 | 670 |
| local 0.75.4 (post-flip, BLAS on) | **3598** | **574** |
| Δ | **−13.2 %** | **−14.3 %** |

Verified BLAS engages via reflection probe: `BlasProvider.IsAvailable = True`, `BackendName = OpenBLAS (AIDOTNET_USE_BLAS=1)`, `AIDOTNET_USE_BLAS env = '(unset)'`.

### Tensors-side block-perf test

`Issue319TransformerBlockPerfTests` (Tensors test project, no OpenBLAS deployed — measures the SimdGemm fallback path):

```
TensorMatMul       :  2480 µs/call (total 893 ms over 360 calls)
TensorBroadcastAdd :   155 µs/call
GELU               :   225 µs/call
LayerNorm          :   738 µs/call    ← regressed from earlier 220 baseline,
                                        likely from one of #321's earlier
                                        commits (not my LayerNorm migration —
                                        verified by revert+retest).
Full block chain   : 51.23 ms/iter
```

12-layer block chain is **51 ms/iter** — down from **630 ms/iter** in the maintainer's pre-PR baseline. That's a **12× block-level speedup** combining:
- BLAS-on-by-default (matmul kernel)
- AutoTracer.ShouldRecord closure-allocation gate (805286de)
- SIMD-vectorize bias-add fast path (3d1e7d94)
- LayerNorm small-batch grain-size migration (6e4c8bb6)

### Why end-to-end is only 14% when block is 12×

The maintainer's perf test measures isolated forward-only block ops at fixed shapes. The consumer ViT-Base train step has additional cost outside that hot path:

- **Backward + optimizer estimate: 3024 ms/iter** (84% of the 3598 ms total) — gradient tape recording, backward op dispatch, parameter buffer alias setup, optimizer Adam step.
- Tape recording overhead per op (closure capture, type-erased delegate boxing).
- Lazy layer initialization on first forward.

BLAS-default-on improves both forward and backward equally (both call `Engine.TensorMatMul`), so the 14% improvement is consistent with backward also seeing 14%. To collapse the remaining gap to 6× consumer-side requires reducing **tape framework overhead** rather than the kernel — that's a separate engineering effort (compiled-training-plan path is the existing infrastructure for it).

### Status

- ✅ BLAS-default-on lands the kernel-level win that `at::parallel_for` grain-size alone couldn't deliver.
- ⏳ ViT-Base `MoreData_ShouldNotDegrade` still times out at 120s × 250 iters — gap is now in the tape framework, not the matmul kernel.
- ⏳ The `Parallel.For` comprehensive migration in PR #321 keeps adding incremental gains for non-matmul ops; happy to keep extending it.

The PR is ready for review at [ooples/AiDotNet.Tensors#321](https://github.com/ooples/AiDotNet.Tensors/pull/321) — three commits from me on top of the maintainer's three commits, all green on net10 and net471.

**Comment by ooples** [comment]
## PR #321 cumulative status — ViT-Base −21.8% train, −25.2% predict

After landing the full chain of #319 work on PR #321 (BLAS-default-on, double fused-optimizer, double specialized matmul forward+backward, LayerNorm grain-size migration, GLU/SwiGLU/ReGLU/GeGLU forward+backward grain-size migrations, ScaledDotProductAttention grain-size migrations) — measured on consumer ViT-Base 224×224×3 train step:

| Stage | Train ms/iter | Predict ms/iter | Δ vs 0.75.3 |
|---|---:|---:|---:|
| Tensors 0.75.3 (baseline, no PR #321) | 4144 | 670 | — |
| + LayerNorm grain-size + BLAS-default-on | 3598 | 574 | −13% / −14% |
| + Double fused optimizer step | 3546 | 540 | −14% / −19% |
| + Double specialized MatMul fwd+bwd | 3451 | 534 | −16.7% / −20.3% |
| + GLU/SwiGLU/SDPA grain-size migrations | **3242** | **501** | **−21.8% / −25.2%** |

Each migration stacks. The grain-size threshold approach (industry-standard `at::parallel_for`-style serial fallback below 32K elementwise work) plus engaging BLAS for double matmul plus the compile-once-replay-many fused training plan for double models all compound.

## What's in PR #321 now (12 commits)

| Commit | What |
|---|---|
| `7c7af94d` | ViT-Base perf-test fixture (maintainer) |
| `97c98d31` | `ParallelForOrSerial` helper + 3 BatchMatMul migrations (maintainer) |
| `3d1e7d94` | SIMD-vectorize bias-add fast path (maintainer) |
| `805286de` | AutoTracer.ShouldRecord closure-allocation gate (maintainer) |
| `8f7157e4` | Diagnostic test (maintainer) |
| `6e4c8bb6` | LayerNorm small-batch grain-size migration |
| `ddbe6d57` | **BLAS default-on** (industry-standard; opt-out via `AIDOTNET_USE_BLAS=0`) |
| `8778263c` | FusedOptimizer Adam/AdamW/SGD **double AVX2 kernels** |
| `38fe6add` | ConfigureOptimizer **double dispatch** |
| `01603808` | **Specialized MatMul forward+backward for double** + `BlasProvider.TryGemmEx(double)` |
| `e4fcb33b` | 11 outerSize / batchSize migrations (GLU family + ND batched matmul) |
| `a8d23fba` | 8 ScaledDotProductAttention forward+backward migrations |

## What's still on the table

**~187 Parallel.For sites remain** in CpuEngine.cs but they're concentrated in:
- Conv2D backward variants (~74 sites) — not on ViT inference path; ViT uses Conv2D once for patch embedding
- Pool variants (~30 sites) — ViT uses no pooling; mainly affects ResNet/EfficientNet families
- BatchNorm variants — already self-threshold-guarded (numChunks self-clamps to 1 for small length)
- Conv3D / video ops — not on transformer family path

For each migration the per-site win on ViT is now <1 ms/iter; further sweep is best driven by other model families (CNN backbones, video models) that hit those op classes. Happy to continue if useful — let me know which family to prioritize.

For the remaining tape-framework cost (~3000 ms/iter backward+optimizer of the 3242 ms/iter total), the leverage point is now:
1. **Phase B `MatMul→ReLU→MatMul` fusion** for double (the third float-only gate at line 2383 in `CompiledTrainingPlan.TryFuseForwardBackward`) — uses `FusedMultiLayerGemm` which is float-only and would need a double equivalent. ViT uses GELU+LayerNorm between matmuls, not ReLU, so this specifically wouldn't help ViT but would help ResNet/MLP families.
2. **Reduce per-op tape recording overhead** in the eager path — `DifferentiableOps.RecordBinary/RecordUnary` allocates closures per op. Could use struct-based op metadata + pool of struct buffers for hot ops.

Both are larger engineering efforts beyond the grain-size migration scope.
